src/file_pipe_log/pipe.rs

```diff
@@ -428,68 +330,105 @@ impl<F: FileSystem> SinglePipe<F> {
     }
 
     fn file_span(&self) -> (FileSeq, FileSeq) {
-        let files = self.files.read();
-        (
-            files.first_seq_in_use,
-            files.first_seq + files.fds.len() as u64 - 1,
-        )
+        let files = self.active_files.read();
+        (files[0].seq, files[files.len() - 1].seq)
     }
 
     fn total_size(&self) -> usize {
-        let files = self.files.read();
-        files.fds.len() * self.target_file_size
+        let (first_seq, last_seq) = self.file_span();
+        (last_seq - first_seq + 1) as usize * self.target_file_size
     }
 
     fn rotate(&self) -> Result<()> {
-        self.rotate_imp(&mut self.active_file.lock())
+        self.rotate_imp(&mut self.writable_file.lock())
     }
 
-    /// Purge obsolete log files to the specific `FileSeq`.
-    ///
-    /// Return the actual removed count of purged files.
     fn purge_to(&self, file_seq: FileSeq) -> Result<usize> {
-        let (prev, current) = self.files.write().logical_purge(file_seq);
-        if file_seq > prev.first_seq + prev.total_len as u64 - 1 {
-            debug_assert_eq!(prev, current);
-            return Err(box_err!("Purge active or newer files"));
-        } else if prev == current {
-            return Ok(0);
-        }
-        for seq in prev.first_seq..current.first_seq {
-            let file_id = FileId {
-                queue: self.queue,
-                seq,
+        let (len, purged_files) = {
+            let mut files = self.active_files.write();
+            let off = (file_seq - files[0].seq) as usize;
```

Comment:
`file_seq < files[0].seq`?

Reply:
Is there a test?

Reply:
Ping @LykxSassinator

Reply:
Ping again.

Reply:
This branch has already been tested in `test_pipe_log`:
```Rust
        ...
        // cannot purge active file
        assert!(pipe_log.purge_to(FileId { queue, seq: 4 }).is_err());
        ...
```
And more boundary test cases have been supplemented into `test_pipe_log_with_recycle`.

Reply:
Let's keep this in mind. Line coverage can only be useful when there's something not covered. What's covered is far from enough and cannot be used as a proof of software quality. (MC/DC coverage, on the other hand, is a much better criterion.) Whenever there's a bug, it's vital to keep the condition covered before fixing it.